src/agent/onefuzz-supervisor/src/commands.rs

```diff
@@ -43,69 +53,72 @@ pub async fn add_ssh_key(key_info: SshKeyInfo) -> Result<()> {
             .await?;
     }
 
-    debug!("removing Authenticated Users permissions from administrators_authorized_keys");
-    let result = Command::new("icacls.exe")
-        .arg(&admin_auth_keys_path)
-        .arg("/remove")
-        .arg("NT AUTHORITY/Authenticated Users")
-        .stdin(Stdio::null())
-        .stdout(Stdio::piped())
-        .stderr(Stdio::piped())
-        .spawn()
-        .context("icacls failed to start")?
-        .wait_with_output()
-        .await
-        .context("icalcs failed to run")?;
-    if !result.status.success() {
-        bail!(
-            "set authorized_keys ({}) permissions failed: {:?}",
-            admin_auth_keys_path.display(),
-            result
-        );
-    }
-
-    debug!("removing inheritance");
-    let result = Command::new("icacls.exe")
-        .arg(&admin_auth_keys_path)
-        .arg("/inheritance:r")
-        .stdin(Stdio::null())
-        .stdout(Stdio::piped())
-        .stderr(Stdio::piped())
-        .spawn()
-        .context("icacls failed to start")?
-        .wait_with_output()
-        .await
-        .context("icacls failed to run")?;
-    if !result.status.success() {
-        bail!(
-            "set authorized_keys ({}) permissions failed: {:?}",
-            admin_auth_keys_path.display(),
-            result
-        );
-    }
-
-    debug!("copying ACL from ssh_host_dsa_key");
-    let result = Command::new("powershell.exe")
-        .args(&["-ExecutionPolicy", "Unrestricted", "-Command"])
-        .arg(format!(
-            "Get-Acl \"{}\" | Set-Acl \"{}\"",
-            admin_auth_keys_path.display(),
-            host_key_path.display()
-        ))
-        .stdin(Stdio::null())
-        .stdout(Stdio::piped())
-        .stderr(Stdio::piped())
-        .spawn()
-        .context("Powershell Get-ACL | Set-ACL failed to start")?
-        .wait_with_output()
-        .await
-        .context("Powershell Get-ACL | Set-ACL failed to run")?;
-    if !result.status.success() {
-        bail!(
-            "set authorized_keys ({}) permissions failed: {:?}",
-            admin_auth_keys_path.display(),
-            result
-        );
+    if SET_PERMISSION_ONCE.set(()).is_ok() {
```

Comment:
I suggest we match on the result of `set()`, and only fall through on `Err(SetErr::AlreadyInitialized)`.

Reply:
In the current impl, is it correct that a caller could attempt this once, succeed with _some_ `icacls` invocations, then fail, and then call this again with a partially-initialized state?

Reply:
I'm wondering if we really need either multiple `OnceCell`'s or an atomics + `lazy_static` state machine.

Reply:
OnceCell is already thread-safe.  I'm not sure why we need atomics on top of this.

Beyond that, we have significant other issues if we handle executing node commands concurrently, so we should not see this executed in anything but a serial fashion.

If any node command fails, we hard fail the supervisor.  As such, we should only ever enter this block once and only once.

Reply:
> I'm not sure why we need atomics on top of this.

Not suggesting atomics are needed for `OnceCell`. Only mentioned them wrt the idea of using a `lazy_static` state machine that required mutation to drive forward (which I'm not necessarily claiming we should do!).

>  so we should not see this executed in anything but a serial fashion.

I understand. My suggestion is that also actually checking this doesn't take too much work (just the `match`, vs. conditional check), so it seems worthwhile. Your call.

> If any node command fails, we hard fail the supervisor.

True, currently. Again, your call, just thinking about future-proofing.

Reply:
Added the other match statements